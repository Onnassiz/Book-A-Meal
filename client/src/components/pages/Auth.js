import React from 'react';
import AuthContainer from '../container/AuthContainer';

const Home = () => (
  <div>
    <AuthContainer />
  </div>
);

export default Home;
