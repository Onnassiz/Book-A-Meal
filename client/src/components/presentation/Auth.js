import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { BasicInput } from '../presentation/form/BasicInput';
import SubmitButton from '../presentation/form/SubmitButton';
import { validateSignUp, validateSignIn } from '../../utilities/validateAuthForms';
import ShowErrors from './partials/ShowErrors';

class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      type: props.match.params.type,
      checked: false,
      isSignIn: false,
      fullName: '',
      signUpEmail: '',
      signInEmail: '',
      signInPassword: '',
      password: '',
      role: 'customer',
      confirm_password: '',
      signUpErrors: {},
      signInErrors: {},
      errors: {},
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleSignInSubmit = this.handleSignInSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onCheckBoxChange = this.onCheckBoxChange.bind(this);
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  onCheckBoxChange(e) {
    const { value } = e.target;
    this.setState({
      checked: !this.state.checked,
      role: value,
    });
  }

  isValid() {
    this.setState({ signUpErrors: {} });
    const { errors, isValid } = validateSignUp(this.state);
    if (!isValid) {
      this.setState({ signUpErrors: errors });
    }
    return isValid;
  }

  isValidSignIn() {
    this.setState({ signInErrors: {} });
    const { errors, isValid } = validateSignIn(this.state);
    if (!isValid) {
      this.setState({ signInErrors: errors });
    }
    return isValid;
  }

  toggleIsSignIn(val) {
    this.setState({ isSignIn: val });
  }

  handleSubmit(e) {
    e.preventDefault();
    this.toggleIsSignIn(false);
    if (this.isValid()) {
      const formData = {
        fullName: this.state.fullName,
        email: this.state.signUpEmail,
        password: this.state.password,
        role: this.state.role,
      };

      const { postUser } = this.props;
      postUser(formData);
    }
  }

  handleSignInSubmit(e) {
    this.toggleIsSignIn(true);
    e.preventDefault();
    if (this.isValidSignIn()) {
      const formData = {
        email: this.state.signInEmail,
        password: this.state.signInPassword,
      };

      const { signInUser } = this.props;
      signInUser(formData);
    }
  }

  changeAuthType(type) {
    const { history } = this.props;
    this.setState({ type });
    history.push(`/auth/${type}`);
  }

  showRegisterLoginLinks() {
    return (
      <div style={{ clear: 'both', marginTop: 50, textAlign: 'center' }}>
        <hr />
        {this.state.type === 'login' ?
          <a id="register" style={{ fontSize: 17 }} onClick={() => this.changeAuthType('register')}><i className="ion-person-add" /> Register</a> :
          <a id="signIn" style={{ fontSize: 17 }} onClick={() => this.changeAuthType('login')}><i className="ion-log-in" /> Login</a>
        }
      </div>
    );
  }

  renderSignUp() {
    const { user, formState } = this.props;
    return (
      <div className="box-auth">
        <h3>Sign Up</h3>
        <ShowErrors
          clientErrors={this.state.signUpErrors}
          serverErrors={user.signUpErrors || {}}
        />
        <form onSubmit={this.handleSubmit}>
          <BasicInput name="fullName" type="text" label="Full Name" value={this.state.fullName} onChange={this.onChange} hasError={this.state.signUpErrors.fullName !== undefined} />
          <BasicInput name="signUpEmail" type="text" label="Email" value={this.state.signUpEmail} onChange={this.onChange} hasError={this.state.signUpErrors.signUpEmail !== undefined} />
          <BasicInput name="password" type="password" label="Password" value={this.state.password} onChange={this.onChange} hasError={this.state.signUpErrors.password !== undefined} />
          <BasicInput name="confirm_password" type="password" label="Confirm Password" value={this.state.confirm_password} onChange={this.onChange} hasError={this.state.signUpErrors.confirm_password !== undefined} />
          <div id="signUpAsAdmin">
            <label className="checkbox" htmlFor="signUp">
            Sign up as Caterer
              <input id="signUp" value={this.state.checked ? 'customer' : 'caterer'} onChange={this.onCheckBoxChange} checked={this.state.checked} type="checkbox" /><span className="check" />
            </label>
          </div>
          <SubmitButton value="Sign Up" isLoading={formState.isLoading && this.state.isSignIn} />
        </form>
      </div>
    );
  }

  renderLogin() {
    const { user, formState } = this.props;
    return (
      <div className="box-auth">
        <h3>Sign In</h3>
        <ShowErrors
          clientErrors={this.state.signInErrors}
          serverErrors={user.signInErrors || {}}
        />
        <form onSubmit={this.handleSignInSubmit}>
          <BasicInput name="signInEmail" type="text" label="Email" value={this.state.signInEmail} onChange={this.onChange} hasError={this.state.signInErrors.signInEmail !== undefined} />
          <BasicInput name="signInPassword" type="password" label="Password" value={this.state.signInPassword} onChange={this.onChange} hasError={this.state.signInErrors.signInPassword !== undefined} />
          <div id="forgotPassword">
            <Link to="/" href="/">Forgot Password?</Link>
          </div>
          <SubmitButton value="Sign In" isLoading={formState.isLoading} />
        </form>
      </div>
    );
  }

  render() {
    return (
      <div>
        <div id="content3">
          <div className="col-12">
            <Link className="main" to="/" href="/"><button className="button-white">Home</button></Link>
            <div id="authSection">
              {this.state.type === 'login' ? this.renderLogin() : this.renderSignUp()}
              {this.showRegisterLoginLinks()}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Home.propTypes = {
  signInUser: PropTypes.func.isRequired,
  postUser: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  formState: PropTypes.object.isRequired,
};


export default Home;
