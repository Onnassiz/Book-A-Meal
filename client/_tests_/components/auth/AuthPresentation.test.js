import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Auth from '../../../src/components/presentation/Auth';

const props = {
  signInUser: jest.fn(),
  postUser: jest.fn(),
  history: {
    push: jest.fn(),
  },
  user: {},
  match: { params: { type: 'login' } },
  formState: { isLoading: false },
};

const wrapper = shallow(<Auth {...props} />);

props.match = { params: { type: 'register' } };
const registerWrapper = shallow(<Auth {...props} />);

jest.useFakeTimers();
describe('<Auth />', () => {
  it('Should render Auth presentation components successfully', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should render Auth presentation components successfully with signUp form', () => {
    expect(toJson(registerWrapper)).toMatchSnapshot();
  });

  it('Should call handle change method to update handle input change', () => {
    const onChangeSpy = jest.spyOn(wrapper.instance(), 'onChange');
    const firstEvent = {
      target: {
        name: 'fullName',
        value: 'Benjamin Onah',
      },
    };
    wrapper.instance().onChange(firstEvent);
    expect(wrapper.state().fullName).toBe('Benjamin Onah');
    expect(onChangeSpy).toHaveBeenCalled();
  });

  it('Should set role to caterer when checkbox is clicked', () => {
    const event = {
      target: { value: 'caterer' },
    };
    registerWrapper.find('#signUpAsCaterer').simulate('change', event);
    expect(registerWrapper.state().checked).toBe(true);
    expect(registerWrapper.state().role).toBe('caterer');
  });

  it('Should set auth type to register when register button is clicked', () => {
    wrapper.find('#register').simulate('click');
    expect(wrapper.state().type).toBe('register');
  });

  it('Should set auth type to login when login button is clicked', () => {
    registerWrapper.find('#signIn').simulate('click');
    expect(registerWrapper.state().type).toBe('login');
  });

  it('Should trigger handleSubmit when register form is submitted', () => {
    const handleSubmitSpy = jest.spyOn(registerWrapper.instance(), 'handleSubmit');
    const postUserSpy = jest.spyOn(registerWrapper.instance().props, 'postUser');
    const event = {
      preventDefault: jest.fn(),
    };
    registerWrapper.setState({
      fullName: 'Ben Onah',
      password: 'password',
      confirm_password: 'password',
      signUpEmail: 'onnassiz@andela.com',
      role: 'customer',
    });
    registerWrapper.instance().handleSubmit(event);
    expect(handleSubmitSpy).toHaveBeenCalled();
    expect(postUserSpy).toHaveBeenCalledTimes(1);
  });

  it('Should not trigger postUser when register form is submitted with validation error', () => {
    const handleSubmitSpy = jest.spyOn(registerWrapper.instance(), 'handleSubmit');
    const postUserSpy = jest.spyOn(registerWrapper.instance().props, 'postUser');
    const event = {
      preventDefault: jest.fn(),
    };
    registerWrapper.setState({
      fullName: 'Ben Onah',
      password: 'password',
      confirm_password: 'password',
      signUpEmail: '',
      role: 'customer',
    });
    registerWrapper.instance().handleSubmit(event);
    expect(registerWrapper.state().signUpErrors.signUpEmail).toBe('The email field is required');
    expect(handleSubmitSpy).toHaveBeenCalled();
    expect(postUserSpy).toHaveBeenCalledTimes(1);
  });

  it('Should return errors when form validation fails on register', () => {
    const handleSubmitSpy = jest.spyOn(registerWrapper.instance(), 'handleSubmit');
    const postUserSpy = jest.spyOn(registerWrapper.instance().props, 'postUser');
    const event = {
      preventDefault: jest.fn(),
    };
    registerWrapper.setState({
      fullName: '',
      password: '',
      confirm_password: '',
      signUpEmail: '',
      role: '',
    });
    registerWrapper.instance().handleSubmit(event);
    expect(registerWrapper.state().signUpErrors.signUpEmail).toBe('The email field is required');
    expect(registerWrapper.state().signUpErrors.fullName).toBe('The full name field is required');
    expect(registerWrapper.state().signUpErrors.password).toBe('The password field is required');
    expect(registerWrapper.state().signUpErrors.confirm_password).toBe('The password confirmation field is required');
    expect(handleSubmitSpy).toHaveBeenCalled();
    expect(postUserSpy).toHaveBeenCalledTimes(1);
  });

  it('Should return errors when form validation fails on register', () => {
    const handleSubmitSpy = jest.spyOn(registerWrapper.instance(), 'handleSubmit');
    const postUserSpy = jest.spyOn(registerWrapper.instance().props, 'postUser');
    const event = {
      preventDefault: jest.fn(),
    };
    registerWrapper.setState({
      fullName: '',
      password: 'one-pass',
      confirm_password: 'pass',
      signUpEmail: '',
      role: '',
    });
    registerWrapper.instance().handleSubmit(event);
    expect(registerWrapper.state().signUpErrors.confirm_password).toBe('Password confirmation must match password');
    expect(handleSubmitSpy).toHaveBeenCalled();
    expect(postUserSpy).toHaveBeenCalledTimes(1);
  });

  it('Should trigger handleSubmit when login form is submitted', () => {
    const handleSignInSubmitSpy = jest.spyOn(wrapper.instance(), 'handleSignInSubmit');
    const signInUserSpy = jest.spyOn(wrapper.instance().props, 'signInUser');
    const event = {
      preventDefault: jest.fn(),
    };
    wrapper.setState({
      signInPassword: 'password-1',
      signInEmail: 'onnassiz@andela.com',
    });
    wrapper.instance().handleSignInSubmit(event);
    expect(handleSignInSubmitSpy).toHaveBeenCalled();
    expect(signInUserSpy).toHaveBeenCalledTimes(1);
  });

  it('Should not trigger postUser when login form is submitted with validation error', () => {
    const handleSignInSubmitSpy = jest.spyOn(wrapper.instance(), 'handleSignInSubmit');
    const signInUserSpy = jest.spyOn(wrapper.instance().props, 'signInUser');
    const event = {
      preventDefault: jest.fn(),
    };
    wrapper.setState({
      signInPassword: '',
      signInEmail: '',
    });
    wrapper.instance().handleSignInSubmit(event);
    expect(wrapper.state().signInErrors.signInEmail).toBe('The email field is required');
    expect(wrapper.state().signInErrors.signInPassword).toBe('The password field is required');
    expect(handleSignInSubmitSpy).toHaveBeenCalled();
    expect(signInUserSpy).toHaveBeenCalledTimes(1);
  });
});
