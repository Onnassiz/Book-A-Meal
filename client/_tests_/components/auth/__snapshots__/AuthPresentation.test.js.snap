// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Auth /> Should render Auth presentation components successfully 1`] = `
<div>
  <div
    id="content3"
  >
    <div
      className="col-12"
    >
      <Link
        className="main"
        href="/"
        replace={false}
        to="/"
      >
        <button
          className="button-white"
        >
          Home
        </button>
      </Link>
      <div
        id="authSection"
      >
        <div
          className="box-auth"
        >
          <h3>
            Sign In
          </h3>
          <ShowErrors
            clientErrors={Object {}}
            serverErrors={Object {}}
          />
          <form
            onSubmit={[Function]}
          >
            <BasicInput
              hasError={false}
              label="Email"
              min=""
              name="signInEmail"
              onChange={[Function]}
              type="text"
              value=""
            />
            <BasicInput
              hasError={false}
              label="Password"
              min=""
              name="signInPassword"
              onChange={[Function]}
              type="password"
              value=""
            />
            <div
              id="forgotPassword"
            >
              <Link
                href="/"
                replace={false}
                to="/"
              >
                Forgot Password?
              </Link>
            </div>
            <SubmitButton
              isLoading={false}
              value="Sign In"
            />
          </form>
        </div>
        <div
          style={
            Object {
              "clear": "both",
              "marginTop": 50,
              "textAlign": "center",
            }
          }
        >
          <hr />
          <a
            id="register"
            onClick={[Function]}
            style={
              Object {
                "fontSize": 17,
              }
            }
          >
            <i
              className="ion-person-add"
            />
             Register
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Auth /> Should render Auth presentation components successfully with signUp form 1`] = `
<div>
  <div
    id="content3"
  >
    <div
      className="col-12"
    >
      <Link
        className="main"
        href="/"
        replace={false}
        to="/"
      >
        <button
          className="button-white"
        >
          Home
        </button>
      </Link>
      <div
        id="authSection"
      >
        <div
          className="box-auth"
        >
          <h3>
            Sign Up
          </h3>
          <ShowErrors
            clientErrors={Object {}}
            serverErrors={Object {}}
          />
          <form
            onSubmit={[Function]}
          >
            <BasicInput
              hasError={false}
              label="Full Name"
              min=""
              name="fullName"
              onChange={[Function]}
              type="text"
              value=""
            />
            <BasicInput
              hasError={false}
              label="Email"
              min=""
              name="signUpEmail"
              onChange={[Function]}
              type="text"
              value=""
            />
            <BasicInput
              hasError={false}
              label="Password"
              min=""
              name="password"
              onChange={[Function]}
              type="password"
              value=""
            />
            <BasicInput
              hasError={false}
              label="Confirm Password"
              min=""
              name="confirm_password"
              onChange={[Function]}
              type="password"
              value=""
            />
            <div
              id="signUpAsAdmin"
            >
              <label
                className="checkbox"
                htmlFor="signUp"
              >
                Sign up as Caterer
                <input
                  checked={false}
                  id="signUp"
                  onChange={[Function]}
                  type="checkbox"
                  value="caterer"
                />
                <span
                  className="check"
                />
              </label>
            </div>
            <SubmitButton
              isLoading={false}
              value="Sign Up"
            />
          </form>
        </div>
        <div
          style={
            Object {
              "clear": "both",
              "marginTop": 50,
              "textAlign": "center",
            }
          }
        >
          <hr />
          <a
            id="signIn"
            onClick={[Function]}
            style={
              Object {
                "fontSize": 17,
              }
            }
          >
            <i
              className="ion-log-in"
            />
             Login
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
